op.c: Further optimisations of foreach+indexed to apply to LIST as well

leonerd opened the pull request:
This builds on the work of commit a8394b4e8c, expanding it to apply to any `foreach` loop iterating over any kind of list, not just an in-place array.

* This set of changes requires a perldelta entry, and it is included.

leonerd commented:
Anyone any thoughts? Would be nice to get this in too before next release.

tonycoz commented on op.c:
```diff
@@ -9804,45 +9815,59 @@ Perl_newFOROP(pTHX_ I32 flags, OP *sv, OP *expr, OP *block, OP *cont)
-            iterpflags |= OPpITER_INDEXED;
+            OpLASTSIB_set(first, NULL);
+            expr = first;
         }
-        else
-            goto expr_not_special;
+        else {
+            /* the op_targ slot contained the "was" op_type for an
+             * OP_NULL; clear it or op_free() will get very confused */
+            args->op_targ = 0;
+            OpTYPE_set(args, OP_LIST);
+            OpLASTSIB_set(args, NULL);
```
`OpLASTSIB_set(first, NULL);`
`OpLASTSIB_set(args, NULL);`

Doesn't this break the links back to the parent? (ie. I think it should be `args` rather than `NULL` for the first one).

nm, I think I'm wrong.

leonerd replied:
It doesn't matter too much here. Yes at this point its parent is set to NULL, but in each case this has become the new value for the `expr` tree, which gets given to `op_append_elem()` down at line 9913 or so. That'll set the correct parent when it builds the eventual result structure anyway.

I could add a comment to that effect.

tonycoz replied:
It's mostly that I'm bad at op trees.

leonerd committed:
op.c: Further optimisations of foreach+indexed to apply to LIST as well

This builds on the work of commit a8394b4e8c, expanding it to apply to
any `foreach` loop iterating over any kind of list, not just an in-place
array.
